ydb/core/kqp/opt/physical/effects/kqp_opt_phy_upsert_index.cpp

```diff
@@ -542,13 +540,178 @@ TMaybeNode<TExprList> KqpPhyUpsertIndexEffectsImpl(TKqpPhyUpsertIndexMode mode,
+
+            payloadTuples.emplace_back(
+                Build<TCoNameValueTuple>(ctx, pos)
+                    .Name(columnAtom)
+                    .Value<TCoMember>()
+                        .Struct<TCoNth>()
+                            .Tuple(payloadSelectorArg)
+                            .Index().Build(1)
+                            .Build()
+                        .Name(columnAtom)
+                        .Build()
+                    .Done());
+        }
+
+        auto inputArg = TCoArgument(ctx.NewArgument(pos, "recalc_input_arg_" + indexDesc->Name));
+
+        auto cmp = ctx.Builder(pos)
+            .Callable("AggrNotEquals")
+                .Add(0, Build<TCoAsStruct>(ctx, pos)
+                    .Add(inputRowsForIndex)
+                    .Done().Ptr())
+                .Add(1, Build<TCoAsStruct>(ctx, pos)
+                    .Add(lookupRow)
+                    .Done().Ptr())
+            .Seal().Build();
+
+        for (const auto& key : pk) {
+            auto tuple = Build<TCoNameValueTuple>(ctx, pos)
+                .Name().Build(key)
+                .Value<TCoMember>()
+                    .Struct(inputItem)
+                    .Name().Build(key)
+                    .Build()
+                .Done();
+
+            keyTuples.emplace_back(tuple);
+        }
+
+        auto lookupDictArg = TCoArgument(ctx.NewArgument(pos, "recalc_dict_arg_" + indexDesc->Name));
+        auto reComputeDictStage = Build<TDqStage>(ctx, pos)
+            .Inputs()
+                .Add(inputRowsAndKeys.RowsPrecompute) // input rows
+                .Add(lookupDict.Cast()) // dict contains loockuped from table rows
+                .Build()
+            .Program()
+                .Args({inputArg, lookupDictArg})
+                .Body<TCoIterator>()
+                    .List<TCoMap>()
+                    .Input<TCoToList>()
+                        .Optional<TCoJust>()
+                            .Input(lookupDictArg)
+                            .Build()
+                        .Build()
+                    .Lambda()
+                        .Args({"collection"})
+                        .Body<TCoToDict>()
+                            .List<TCoFlatMap>()
+                                .Input(inputArg)
+                                .Lambda()
+                                    .Args({inputItem})
+                                    .Body<TCoMap>()
+                                        .Input<TCoLookup>()
+                                            .Collection("collection")
+                                            .Lookup<TCoAsStruct>()
+                                                .Add(keyTuples)
+                                                .Build()
+                                            .Build()
+                                        .Lambda<TCoLambda>()
+                                            .Args({payload_table_row})
+                                            .Body<TExprList>()
```

Comment:
Maybe put a comment here, something about that this code actually creates a tuple and what the values in it mean.